### PR TITLE
Add Jython and Groovy ScriptEngineFactories

### LINKED ISSUE #521: [Automation - Feature Request] Add Jython and Groovy scripted Actions to Paper UI
The drop down currently only has Javascript.

### PR DESCRIPTION
**Pull request by openhab-5iver**
... and add support for Jython and Groovy scripted Actions and Conditions in Paper UI.

Fixes #521 

Signed-off-by: Scott Rushworth openhab@5iver.com (github: openhab-5iver)

### CODE REVIEW

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/GroovyScriptEngineFactory.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright (c) 2019 Contributors to the Eclipse Foundation
```
Comment (kaikreuzer) [inline bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/GroovyScriptEngineFactory.java]: Please use the license header from https://github.com/openhab/openhab-core/blob/master/licenses/epl-2.0/header.txt

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/GroovyScriptEngineFactory.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import javax.script.ScriptEngine;
+import javax.script.ScriptEngineManager;
+
+import org.openhab.core.automation.module.script.ScriptEngineFactory;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *
```
Comment (kaikreuzer) [inline bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/GroovyScriptEngineFactory.java]: Please add some Javadoc

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/JythonScriptEngineFactory.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import javax.script.ScriptEngine;
+import javax.script.ScriptEngineManager;
+
+import org.openhab.core.automation.module.script.ScriptEngineFactory;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *
```
Comment (kaikreuzer) [inline bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/JythonScriptEngineFactory.java]: Please add some Javadoc

**Comment by openhab-5iver** [comment]
> I'd thus think that the registration of the factories + config description should be constructed dynamically, wdyt?

Absolutely, but I did not see how to dynamically construct it. Could you please point me in the right direction on how to do that?

**Comment by kaikreuzer** [comment]
Well, the first thing is to figure out how to determine whether support of these languages is there or not.
Then you could write some code that registers a marker service and have a mandatory dependency of the factory to that service.

For the config options, this should be removed from the static json file and be implemented as an `ConfigOptionProvider` instead.

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/GroovyScriptEngineFactory.java (hunk 1 above)
Comment (openhab-5iver) [reply bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/GroovyScriptEngineFactory.java]: I'm not sure if I should use this...

    Copyright (c) 2010-${year} Contributors to the openHAB project

Or...

    Copyright (c) 2010-2019 Contributors to the openHAB project

Comment (kaikreuzer) [reply bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/GroovyScriptEngineFactory.java]: Clearly the version with the years, not the placeholders!

**Comment by kaikreuzer** [comment]
I think 2+3 should be combined.

> Adding some bundles for JSR223 libraries is on my list though! What type of bundles would these be?

What do you mean by type? Wrt the id, it could be `org.openhab.core.automation.module.script.jython` and `org.openhab.core.automation.module.script.groovy`.

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/GroovyScriptEngineFactory.java (hunk 1 above)
Comment (openhab-5iver) [reply bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/GroovyScriptEngineFactory.java]: I thought so... but clearly not so clear to me! Thank you! :smile:

**Comment by openhab-5iver** [comment]
> I think 2+3 should be combined.

If 3 is implemented, I don't see any need for 2. The bundle could be installed but misconfigured. Checking the classpath will tell us for certain if the libraries are available for use.

> What do you mean by type?

I was meaning binding, io, etc., but don't know the right terminology for it.

I have a number of questions, so I opened a separate issue for Jython... https://github.com/openhab/openhab2-addons/issues/4801, and I'll open another for Groovy.

**Comment by openhab-5iver** [comment]
Replaced with #635.

**Pull request closed without merging**